text/0049-metric-label-set.md

```diff
@@ -0,0 +1,105 @@
+
+## Explanation
+
+Metric instrument APIs which presently take labels in the form `{ Key: Value, ... }` will be updated to take an explicit `LabelSet`.  The `Meter.Labels()` API method supports getting a `LabelSet` from the API, allowing the programmer to acquire a pre-defined label set.  Here are several examples of `LabelSet` re-use.  Assume we have two instruments:
+
+```golang
+var (
+    cumulative = metric.NewFloat64Cumulative("my_counter")
+    gauge      = metric.NewFloat64Gauge("my_gauge")
+)    
+```
+
+Use a `LabelSet` to construct multiple Handles:
+
+```golang
+var (
+    labels  = meter.Labels({ "required_key1": value1, "required_key2": value2 })
+    chandle = cumulative.GetHandle(labels)
+    ghandle = gauge.GetHandle(labels)
+)
+for ... {
+   // ...
+   chandle.Add(...)
+   ghandle.Set(...)
+}
+```
+
+Use a `LabelSet` to make multiple Direct calls:
+
+```golang
+labels := meter.Labels({ "required_key1": value1, "required_key2": value2 })
+cumulative.Add(quantity, labels)
+gauge.Set(quantity, labels)
+```
+
+Of course, repeated calls to `Meter.RecordBatch()` could re-use a `LabelSet` as well.
+
+### Ordered `LabelSet` option
+
+As a language-level decision, APIs may support _ordered_ LabelSet
+construction, in which a pre-defined set of ordered label keys is
+defined such that values can be supplied in order.  This allows a
+faster code path to construct the `LabelSet`.  For example,
+
+```golang
+
+var rpcLabelKeys = meter.OrderedLabelKeys("a", "b", "c")
+
+for _, input := range stream {
+    labels := rpcLabelKeys.Values(1, 2, 3)  // a=1, b=2, c=3
+
+    // ...
+}
+```
+
+This is specified as a language-optional feature because its safety,
+and therefore its value as an input for monitoring, depends on the
+availability of type-checking in the source language.  Passing
+unordered labels (i.e., a list of bound keys and values) to
+`Meter.Labels(...)` is considered the safer alternative.
+
+## Internal details
+
+Metric SDKs that do not or cannot take advantage of the LabelSet optimizations are not especially burdened by having to support these APIs.  It is trivial to supply an implementation of `LabelSet` that simply stores a list of labels.  This may not be acceptable in performance-critical applications, but this is the common case in many metrics and diagnostics APIs today.
+
+## Trade-offs and mitigations
+
+In languages where overloading is a standard convenience, the metrics API may elect to offer alternate forms that elide the call to `Meter.Labels()`, for example:
+
+```
+instrument.GetHandle(meter, { Key: Value, ... })
+```
+
+as opposed to this:
+
+```
+instrument.GetHandle(meter.Labels({ Key: Value, ... }))
```

Comment:
Can I simply have a version:
```
instrument.GetHandle({ Key: Value, ... })
```

Reply:
This is already stated in "Tradeoffs and Mitigations" below and appears in the (current spec)[https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics-user.md#option-convenience-method-to-bypass-meterlabels].

Reply:
After removing the meter correct :). I missed this.